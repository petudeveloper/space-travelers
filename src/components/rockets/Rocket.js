import { PropTypes } from 'prop-types';
import { Card, Button } from 'react-bootstrap';
import { useDispatch } from 'react-redux';
import style from './rocket.module.css';
import { reserveRocket, cancelRocket } from '../../redux/rockets/rockets';

const Rocket = ({
  name, imageUrl, description, id, reserved,
}) => {
  const dispatch = useDispatch();

  const reserveRocketHandler = () => {
    if (reserved) dispatch(cancelRocket(id));
    else dispatch(reserveRocket(id));
  };

  return (
    <li className={style.cardContainer}>
      <Card className={style.cardStyle}>
        <Card.Img
          variant="left"
          src={imageUrl}
          className={style.rocketImage}
        />
        <Card.Body>
          <Card.Title className={style.rocketTitle}>{name}</Card.Title>
          <Card.Text>
            {description}
          </Card.Text>
          <Button
            onClick={reserveRocketHandler}
            variant={reserved ? 'secondary' : 'primary'}
          >
            {reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
          </Button>
        </Card.Body>
      </Card>
    </li>
  );
};

Rocket.propTypes = {
  name: PropTypes.string.isRequired,
  imageUrl: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};

export default Rocket;
