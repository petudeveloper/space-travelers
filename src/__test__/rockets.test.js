import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/cofigureStore';
import Rocket from '../components/rockets/Rocket';
import '@testing-library/jest-dom';
import RocketsList from '../components/rockets/ListOfRockets';

beforeEach(() => {
  render(
    <Provider store={store}>
      <Rocket
        key="Rocket Key Test"
        name="Rocket Test"
        imageUrl="Rocket Test ImageURL"
        description="Rocket Test Description."
        id={1}
        reserved
      />
    </Provider>,
  );
});
afterEach(cleanup);

describe('List of Rockets Tests', () => {
  test('RocketsList should display properly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <RocketsList />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Rockets Tests', () => {
  test('Reserve badge should appear when a Rocket is reserved', () => {
    screen.getByText('Rocket Test');
    expect(screen.getByText('Reserved')).toBeInTheDocument();
  });
  test('Reserve button should appear when a Rocket is reserved', () => {
    screen.getByText('Rocket Test');
    expect(screen.getByText('Cancel Reservation')).toBeInTheDocument();
  });
});
