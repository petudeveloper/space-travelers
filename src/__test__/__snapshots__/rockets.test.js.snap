// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List of Rockets Tests RocketsList should display properly 1`] = `<ul />`;
